src/vscode-bicep/src/panes/deploy/app/components/hooks/useMessageHandler.ts

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+import { useEffect, useState } from "react";
+import { vscode } from "../../vscode";
+import {
+  VscodeMessage,
+  createGetAccessTokenMessage,
+  createGetDeploymentScopeMessage,
+  createGetStateMessage,
+  createPickParamsFileMessage,
+  createReadyMessage,
+  createSaveStateMessage,
+  createShowUserErrorDialogMessage,
+} from "../../../messages";
+import { parseParametersJson, parseTemplateJson } from "../utils";
+import {
+  DeployPaneState,
+  DeploymentScope,
+  ParametersMetadata,
+  TemplateMetadata,
+  UntypedError,
+} from "../models";
+import { AccessToken } from "@azure/identity";
+
+// TODO see if there's a way to use react hooks instead of this hackery
+let accessTokenResolver: {
+  resolve: (accessToken: AccessToken) => void;
+  reject: (error: UntypedError) => void;
+};
```

Comment:
There's got to be a better way to do this...

Reply:
We may create a promise in App.tsx with `useRef``, pass that to useMessageHandler and pass () => promise to useAzure. This can be implement in a future PR though.

Reply:
A related question - would it make sense to move the ARM API calls to the language server (or even Bicep.Core since we've been taking about implementing `bicep deploy`)? That will make the client implementation much simpler.